### PR TITLE
feat: add Training Hub callbacks example for OpenShift AI

### PR DESCRIPTION
**Pull request by Sridhar1030**
- Add `examples/fine-tuning/callbacks/` with a README and end-to-end
  cluster smoke notebook demonstrating unified Training Hub callbacks via the
  Kubeflow SDK `TrainingHubTrainer(callbacks=[...])` API.
- Covers all three backends: Unsloth (lora_sft), InstructLab Training (sft),
  and Mini-Trainer (osft).
- Before/after comparison (6 TrainJobs) proves callbacks inject and fire
  without backend-specific rewrites.
## Related
- Kubeflow SDK PR: https://github.com/opendatahub-io/kubeflow-sdk/pull/137
- Training Hub PR: https://github.com/Red-Hat-AI-Innovation-Team/training_hub/pull/128 , https://github.com/Red-Hat-AI-Innovation-Team/training_hub/pull/141
## Test plan
- [x] Notebook runs on OpenShift AI cluster with `training-hub` runtime
- [x] All 6 jobs (3 baselines + 3 with callbacks) produce expected log markers

### CODE REVIEW

File: examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "0b9dd39f",
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "# Install the SDK with callbacks support.\n",
+        "%pip install -q \"git+https://github.com/opendatahub-io/kubeflow-sdk.git@feat/rhoaieng-79848-sdk-callbacks\"\n",
```
[the remark this answers is not in the document]
Comment (Sridhar1030) [reply examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb:42]: Intentional Branch tips are used on purpose until stable SDK and Training Hub releases ship. Added notes in cell1 and cell 2 of the notebook; we'll pin to release tags when those versions are available.

File: examples/fine-tuning/callbacks/README.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
+# Training Hub Callbacks on Red Hat OpenShift AI
+
+This example demonstrates using unified Training Hub callbacks when submitting TrainJobs with `TrainingHubTrainer` from the Kubeflow SDK on Red Hat OpenShift AI.
+
+## Note
+
+This example is compatible with RHOAI version 3.3+ and requires a Training Hub build with unified callback support.
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/README.md]: This feature is going GA in RHOAI 3.6.EA1 version
Is it compatible with previous versions too ?

Comment (Sridhar1030) [reply examples/fine-tuning/callbacks/README.md]: I'll update the README from "3.3+" to "3.6.EA1+", Thanks for catching that.

File: examples/fine-tuning/callbacks/README.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
+# Training Hub Callbacks on Red Hat OpenShift AI
+
+This example demonstrates using unified Training Hub callbacks when submitting TrainJobs with `TrainingHubTrainer` from the Kubeflow SDK on Red Hat OpenShift AI.
+
+## Note
+
+This example is compatible with RHOAI **3.6.EA1+** only. Unified Training Hub callbacks are not available on earlier RHOAI releases.
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/README.md]:
```suggestion
This example is compatible with RHOAI **3.6.EA1+**. Unified Training Hub callbacks are not available on earlier RHOAI releases.
```

**Comment by ChughShilpa** [comment]
Also add link of this new example in `Learn more` section of https://github.com/red-hat-data-services/red-hat-ai-examples/tree/main/examples/fine-tuning

File: examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+        "Uses the RHOAIENG-79848 feature branch until callbacks land in a stable SDK release."
+      ],
+      "id": "7fb27b941602401d91542211134fc71a"
+    },
+    {
+      "cell_type": "code",
+      "metadata": {},
+      "source": [
+        "# Install the SDK with callbacks support.\n",
+        "%pip install -q \"git+https://github.com/opendatahub-io/kubeflow-sdk.git@feat/rhoaieng-79848-sdk-callbacks\"\n",
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb:42]: Make sure to update it

File: examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb [hunk 5]
```diff
@@ -0,0 +1,802 @@
+{
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "# Training Hub Unified Callbacks — Example Notebook\n",
+        "\n",
+        "**Jira:** RHOAIENG-79848 (SDK) · **Epic:** RHAISTRAT-1256 (unified callbacks)\n",
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb]: Remove this line

File: examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb [hunk 6]
```diff
@@ -0,0 +1,802 @@
...
+      ],
+      "id": "0334bf02"
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## 1. Install SDK branch (run once per environment)\n",
+        "\n",
+        "Uses the RHOAIENG-79848 feature branch until callbacks land in a stable SDK release."
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb]: Remove this too

File: examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb [hunk 7]
```diff
@@ -0,0 +1,802 @@
...
+        "# Install the SDK with callbacks support.\n",
+        "%pip install -q \"git+https://github.com/opendatahub-io/kubeflow-sdk.git@feat/rhoaieng-79848-sdk-callbacks\"\n",
+        "\n",
+        "import inspect\n",
+        "\n",
+        "from kubeflow.trainer import TrainerClient\n",
+        "from kubeflow.trainer.rhai import TrainingHubAlgorithms, TrainingHubTrainer\n",
+        "\n",
+        "assert \"callbacks\" in inspect.signature(TrainingHubTrainer).parameters, (\n",
+        "    \"TrainingHubTrainer.callbacks missing — reinstall the RHOAIENG-79848 SDK branch\"\n",
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb]: Remove this reference here too

File: examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb [hunk 8]
```diff
@@ -0,0 +1,802 @@
...
+        "    (\"Unsloth\", \"lora_sft\", \"Baseline (no callbacks=)\", \"unsloth_before\"),\n",
+        "    (\"Unsloth\", \"lora_sft\", \"With callbacks (UnslothSmokeLogger)\", \"unsloth_after\"),\n",
+        "    (\"InstructLab\", \"sft\", \"Baseline (no callbacks=)\", \"sft_before\"),\n",
+        "    (\"InstructLab\", \"sft\", \"With callbacks (InstructLabSmokeLogger)\", \"sft_after\"),\n",
+        "    (\"MiniTrainer\", \"osft\", \"Baseline (no callbacks=)\", \"osft_before\"),\n",
+        "    (\"MiniTrainer\", \"osft\", \"With callbacks (MiniTrainerSmokeLogger)\", \"osft_after\"),\n",
+        "]\n",
+        "\n",
+        "print(\"\\n\" + \"=\" * 72)\n",
+        "print(\"Verification summary — Unified Training Hub Callbacks (RHOAIENG-79848)\")\n",
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb]: Remove jira reference

File: examples/fine-tuning/callbacks/README.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+### 2. Submit a TrainJob
+
+```python
+from kubeflow.trainer import TrainerClient
+from kubeflow.trainer.rhai import TrainingHubAlgorithms, TrainingHubTrainer
+
+from my_callbacks import LossLogger
+
+trainer = TrainingHubTrainer(
+    algorithm=TrainingHubAlgorithms.LORA_SFT,
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/README.md:95]: LORA_SFT can look like the callback is LoRA-specific. A short comment (or one-liner under the snippet) would help, e.g. same LossLogger works with SFT / OSFT / LORA_SFT.

```
# Any supported algorithm works with the same callback class:
    # SFT, OSFT, LORA_SFT and LORA_GRPO
```

File: examples/fine-tuning/callbacks/README.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+```bash
+jupyter notebook training_hub_callbacks_smoke.ipynb
+```
+
+## Troubleshooting
+
+| Symptom | Fix |
+|---------|-----|
+| `callbacks` not on `TrainingHubTrainer` | Upgrade SDK (RHOAIENG-79848) |
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/README.md]: Remove jira reference

File: examples/fine-tuning/callbacks/README.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+|---------|-------|----------|----------|-----|--------|
+| Unsloth (LoRA SFT) | 1 | 1 | NVIDIA L40/L40S or equivalent | 4 cores | 16Gi |
+| InstructLab (SFT) | 1 | 1 | NVIDIA L40/L40S or equivalent | 4 cores | 16Gi |
+| Mini-Trainer (OSFT) | 1 | 1 | NVIDIA L40/L40S or equivalent | 4 cores | 16Gi |
+
+## Prerequisites
+
+- OpenShift cluster with OpenShift AI (**RHOAI 3.6.EA1+**) and the **trainer** component enabled
+- `training-hub` ClusterTrainingRuntime available in the cluster
+- Kubeflow SDK with `callbacks=` support ([RHOAIENG-79848](https://redhat.atlassian.net/browse/RHOAIENG-79848))
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/README.md]: Remove jira reference

File: examples/fine-tuning/callbacks/README.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+| Symptom | Fix |
+|---------|-----|
+| `callbacks` not on `TrainingHubTrainer` | Upgrade SDK (RHOAIENG-79848) |
+| SDK markers OK, no hook output | Verify `training_hub` version has callbacks |
+| OSFT crashes `is_main_process` | Pin `mini_trainer@main` in `packages_to_install` |
+| `inspect.getsource` error | Move callback to a `.py` file |
+| `defines unsupported hooks` | Remove non-unified hook methods |
+| `must use a no-argument constructor` | Remove required `__init__` params |
+
+## Links
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/README.md:165]: Remove all these links

File: examples/fine-tuning/callbacks/README.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+trainer = TrainingHubTrainer(
+    algorithm=TrainingHubAlgorithms.LORA_SFT,
+    func_args={
+        "model_path": "Qwen/Qwen2.5-0.5B-Instruct",
+        "data_path": "/data/train.jsonl",
+        "ckpt_output_dir": "/tmp/out",
+        "num_epochs": 1,
+    },
+    callbacks=[LossLogger],
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/README.md:102]: Some readers may assume the callback is tied to `LORA_SFT` algorithm so can add a comment

```
# Same `LossLogger` works with SFT / OSFT / LORA_SFT/LORA_GRPO
```

File: examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb [hunk 14]
```diff
@@ -0,0 +1,802 @@
...
+        "\n",
+        "    if expect_callbacks:\n",
+        "        passed = has_sdk and has_hooks\n",
+        "        print(\n",
+        "            f\"\\nRESULT: {'✅ WITH CALLBACKS OK' if passed else '❌ WITH CALLBACKS FAILED'}\"\n",
+        "        )\n",
+        "    else:\n",
+        "        passed = not has_sdk and not any(marker in logs for marker in callback_markers)\n",
+        "        print(\n",
+        "            f\"\\nRESULT: {'✅ BASELINE OK (no callbacks)' if passed else '❌ BASELINE UNEXPECTED MARKERS'}\"\n",
```
Comment (ChughShilpa) [inline examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb:374]: Baseline can false-PASS when the job never completes.
wait_and_get_logs catches timeout/failure, still returns logs, and baseline verify_logs treats “no markers” as success. A crashed/timed-out baseline with empty/error logs then shows
Add step for job completion before PASS

File: examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb (hunk 4 above)
Comment (Sridhar1030) [reply examples/fine-tuning/callbacks/training_hub_callbacks_smoke.ipynb:42]: Adding a TODO comment above the install line noting it will be replaced with a stable SDK release.

File: examples/fine-tuning/rhoai-3.6/callbacks/README.md
Comment (efazal) [inline examples/fine-tuning/rhoai-3.6/callbacks/README.md:1]: Please move this callbacks folder into `rhoai-3.6` (create one if it doesn't exist) folder

File: examples/fine-tuning/callbacks/README.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+
+This example demonstrates using unified Training Hub callbacks when submitting TrainJobs with `TrainingHubTrainer` from the Kubeflow SDK on Red Hat OpenShift AI.
+
+## Note
+
+This example is compatible with RHOAI **3.6.EA1+**. Unified Training Hub callbacks are not available on earlier RHOAI releases.
+
+## Overview
+
+Training Hub provides a unified callback interface (`TrainingHubCallback`) that works across InstructLab Training, Mini-Trainer, and Unsloth backends. The Kubeflow SDK exposes this through `TrainingHubTrainer(..., callbacks=[MyCallback])`.
```
Comment (efazal) [inline examples/fine-tuning/callbacks/README.md]: Reference `Training Hub`, `InstructLab`,  `Mini-Trainer`, and `Unsloth` repos here.

File: examples/fine-tuning/rhoai-3.6/callbacks/README.md [hunk 16]
```diff
@@ -0,0 +1,170 @@
+# Training Hub Callbacks on Red Hat OpenShift AI
+
+This example demonstrates using unified Training Hub callbacks when submitting TrainJobs with `TrainingHubTrainer` from the Kubeflow SDK on Red Hat OpenShift AI.
+
+## Note
+
+This example is compatible with RHOAI **3.6.EA1+**. Unified Training Hub callbacks are not available on earlier RHOAI releases.
+
```
Comment (efazal) [inline examples/fine-tuning/rhoai-3.6/callbacks/README.md:8]: Maybe add a section to explain 'What is a Callback'

**Review by efazal — COMMENTED** [review]
Thanks @Sridhar1030. Left a few comments.

File: examples/fine-tuning/rhoai-3.6/callbacks/training_hub_callbacks.ipynb
Comment (efazal) [inline examples/fine-tuning/rhoai-3.6/callbacks/training_hub_callbacks.ipynb:1]: can we rename the file to `training_hub_callbacks.ipynb` without "smoke" appended to it.

File: examples/fine-tuning/rhoai-3.6/callbacks/README.md [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    def on_log(self, context: TrainingHubContext) -> None:
+        print(f"step={context.step} loss={context.loss} lr={context.learning_rate}", flush=True)
+
+    def on_train_end(self, context: TrainingHubContext) -> None:
+        print(f"Training finished — step={context.step}", flush=True)
+```
+
+> Callbacks **must** be in a `.py` file (not inline in a notebook). The SDK uses `inspect.getsource()`.
+
```
Comment (efazal) [inline examples/fine-tuning/rhoai-3.6/callbacks/README.md:70]: This is confusing, why do we say it must be in a `.py` file? it doesn't have to be right? have you tested defining the callbacks inline, in a cell?

File: examples/fine-tuning/rhoai-3.6/callbacks/training_hub_callbacks_smoke.ipynb
Comment (efazal) [reply examples/fine-tuning/rhoai-3.6/callbacks/training_hub_callbacks_smoke.ipynb]: Can we also limit it to a single example, since it's the same process for other backends. In the readme you've mentioned support being added 3 backends, and you can say this example is for unsloth, and the same steps can be followed for other backends? so it's easier for the user to follow. As it stands currently, it's quite difficult to follow.

File: examples/fine-tuning/rhoai-3.6/callbacks/training_hub_callbacks_smoke.ipynb [hunk 18]
```diff
@@ -0,0 +1,812 @@
...
+        "from pathlib import Path\n",
+        "\n",
+        "RUNTIME_NAME = os.environ.get(\"TRAINING_RUNTIME\", \"training-hub\")\n",
+        "TRAINING_HUB_REF = os.environ.get(\n",
+        "    \"TRAINING_HUB_REF\",\n",
+        "    \"git+https://github.com/Red-Hat-AI-Innovation-Team/training_hub.git@main\",\n",
+        ")\n",
+        "MINI_TRAINER_REF = os.environ.get(\n",
+        "    \"MINI_TRAINER_REF\",\n",
+        "    \"git+https://github.com/Red-Hat-AI-Innovation-Team/mini_trainer.git@main\",\n",
```
Comment (efazal) [inline examples/fine-tuning/rhoai-3.6/callbacks/training_hub_callbacks_smoke.ipynb]: Before this PR gets merged lets make sure above pip install references are removed

File: examples/fine-tuning/rhoai-3.6/callbacks/README.md (hunk 17 above)
Comment (Sridhar1030) [reply examples/fine-tuning/rhoai-3.6/callbacks/README.md:70]: it doesn't have to be a standalone .py file. The SDK needs importable source for inspect.getsource(), so a regular module or %%writefile in the notebook both work. We haven't tested defining callbacks inline in a cell only; this example uses %%writefile to hub_callback_smoke.py. I've updated the README to reflect that.

**Comment by efazal** [comment]
@Sridhar1030 one little comment, the notebook cells have little to no comments. Would be nice to have some description, not too much, looks a little bare otherwise. Check out other examples for inspiration.

**Review by efazal — APPROVED** [review]
Thanks @Sridhar1030. 
/lgtm
